### ISSUE TITLE
routes not support underline as URI pattern

### ISSUE THREAD

**Issue by ohhjulia** [playframework/playframework#10999]
(Note by @mkurz: I changed `http://xxx.com` to `http://localhost:9000`)

### problem

If I write an URI as below

```
GET         /*path/Gx0YLry_RXM.txt                                          controllers.xxx.xxx(path)
```

when I visit http://localhost:9000/Gx0YLry_RXM.txt, I get an 404 error, but http://localhost:9000/123/Gx0YLry_RXM.txt works.
So I change the route to 

```
GET         /*path/Gx0YLryRXM.txt                                           controllers.xxx.xxx(path)
```

http://localhost:9000/Gx0YLryRXM.txt and http://localhost:9000/123/Gx0YLry_RXM.txt both works,

So I think this is a bug when deal with "*path".

### Play Version

2.7

### API

java

### Operating System

imac

**Comment by chipz** [comment]
I tried this in Java API, Play version 2.7.7 with the example app.
```
# An example controller showing how to use dependency injection
GET     /Gx0YLry_RXM.txt                      controllers.CountController.count
```
this came back 200 OK
tried this also in Scala, with the same version 2.7.7. Also got 200 OK.

Do you have any detailed example that we can try? Or is there something I'm missing here?

**Comment by ohhjulia** [comment]
> 



> I tried this in Java API, Play version 2.7.7 with the example app.
> 
> ```
> # An example controller showing how to use dependency injection
> GET     /Gx0YLry_RXM.txt                      controllers.CountController.count
> ```
> 
> this came back 200 OK tried this also in Scala, with the same version 2.7.7. Also got 200 OK.
> 
> Do you have any detailed example that we can try? Or is there something I'm missing here?

this issue is about the "*path" in route, not the underline “_”.
In my example, the "*path" and “_” show in the route at the same time, then there will be 404.
So, change you route to /*path/Gx0YLry_RXM.txt, and visit http://xxx.com/Gx0YLry_RXM.txt, you will get a 404 error.

**Comment by mkurz** [comment]
> If I write an URI as below
> 
> ```
> GET         /*path/Gx0YLry_RXM.txt                                          controllers.xxx.xxx(path)
> ```
> 
> when I visit http://localhost:9000/Gx0YLry_RXM.txt, I get an 404 error, but http://localhost:9000/123/Gx0YLry_RXM.txt works.

Yes. And this is correct. If you look at the error page in dev mode when visiting http://localhost:9000/Gx0YLry_RXM.txt you will see:
<kbd><img width="800" src="https://user-images.githubusercontent.com/644927/154241277-7a8a63b6-f2af-492d-bacd-57e7a6251d6f.png" /></kbd>
You see [Play replaces](https://github.com/playframework/playframework/blob/a4b4198bf6de7bc599f869971abdf86fa16966c5/dev-mode/routes-compiler/src/main/scala/play/routes/compiler/RoutesFileParser.scala#L233-L235) `/*path/Gx0YLry_RXM.txt` with `/$path<.+>/Gx0YLry_RXM.txt`

> So I change the route to
> 
> ```
> GET         /*path/Gx0YLryRXM.txt                                           controllers.xxx.xxx(path)
> ```
> 
> http://localhost:9000/Gx0YLryRXM.txt and http://localhost:9000/123/Gx0YLry_RXM.txt both works,
> 
> So I think this is a bug when deal with "*path".

I can not reproduce this. For me both url give 404, which also is correct (also tried with Play 2.7.9)
Is it possible you got confused?

Also the [Scala](https://www.playframework.com/documentation/2.8.x/ScalaRouting) and [Java routing](https://www.playframework.com/documentation/2.8.x/JavaRouting) docs only contain examples where `*path` is the last part of the uri, so it isn't documented this will work when you add another static part at the end (`.../Gx0YLry_RXM.txt`)

However, I think I found the solution for what you are trying to achieve by using a regex:
```
GET         /$path<(.*[/])?>Gx0YLry_RXM.txt                                 controllers.xxx.xxx(path)
```
- http://localhost:9000/one/two/Gx0YLry_RXM.txt works and `path` will be `one/two/`
- http://localhost:9000/abc/Gx0YLry_RXM.txt works and `path` will be `abc/`
- http://localhost:9000/Gx0YLry_RXM.txt works and path will be empty string
- http://localhost:9000/fooGx0YLry_RXM.txt will be 404 (see the `foo` prefix)

**Comment by mkurz** [comment]
Here is the reproducer I was using: https://github.com/mkurz/playframework-reproducers/blob/main/playframework-10999
Specially the routes file is interesting: https://github.com/mkurz/playframework-reproducers/blob/main/playframework-10999conf/routes
